File: 01-Programming-Languages/Ruby/Algorithms/Graph/bfs.rb
=begin
Problem: Breadth-First Search (BFS)
Description: Traverse a graph level by level using a queue.

Approach:
- Use a queue to process nodes FIFO
- Track visited nodes to avoid cycles

Time Complexity: O(V + E)
Space Complexity: O(V)

Example:
Input:  adjacency list: 0=>[1,2], 1=>[0,3,4], 2=>[0], 3=>[1], 4=>[1], start = 0
Output: [0, 1, 2, 3, 4]
=end

def bfs(graph, start)
  visited = Array.new(graph.length, false)
  queue = [start]
  visited[start] = true
  result = []
  until queue.empty?
    node = queue.shift
    result << node
    graph[node].each do |neighbor|
      unless visited[neighbor]
        visited[neighbor] = true
        queue << neighbor
      end
    end
  end
  result
end

graph = [
  [1, 2],
  [0, 3, 4],
  [0],
  [1],
  [1]
]

puts "BFS starting from 0: #{bfs(graph, 0).join(' ')}"

File: 01-Programming-Languages/TypeScript/Algorithms/Graph/bfs.ts
/*
Problem: BFS (Breadth-First Search)
Description: Traverse a graph from a start node using BFS.

Approach:
- Use a queue to process nodes level by level, track visited set

Time Complexity: O(V + E)
Space Complexity: O(V)

Example:
Input: graph = {0:[1,2],1:[0,3,4],2:[0],3:[1],4:[1]}, start = 0
Output: [0, 1, 2, 3, 4]
*/

type Graph = { [key: number]: number[] };

function bfs(graph: Graph, start: number): number[] {
  const visited = new Set<number>();
  const queue: number[] = [start];
  const result: number[] = [];
  visited.add(start);
  while (queue.length > 0) {
    const node = queue.shift()!;
    result.push(node);
    for (const neighbor of graph[node]) {
      if (!visited.has(neighbor)) {
        visited.add(neighbor);
        queue.push(neighbor);
      }
    }
  }
  return result;
}

const graph: Graph = { 0: [1, 2], 1: [0, 3, 4], 2: [0], 3: [1], 4: [1] };
console.log('BFS from 0:', bfs(graph, 0));
